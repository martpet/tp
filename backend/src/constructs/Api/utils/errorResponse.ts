import { APIGatewayProxyResult } from 'aws-lambda';
import { getReasonPhrase, StatusCodes } from 'http-status-codes';
import { SetRequired } from 'type-fest';

import { EnvName } from '~/types';

type ErrorResponseOptions = WithExposedError | WithoutExposedError;

type BaseOptions = {
  description?: string;
  error?: unknown;
  statusCode?: StatusCodes;
};

type WithExposedError = SetRequired<BaseOptions, 'error'> & {
  exposeErrorMessage: true;
};

type WithoutExposedError = BaseOptions & {
  exposeErrorMessage?: never | false;
};

const exposeErrorEnvs: EnvName[] = ['personal', 'staging'];

export const errorResponse = (
  traceId: string,
  {
    statusCode = StatusCodes.INTERNAL_SERVER_ERROR,
    description,
    error,
    exposeErrorMessage = exposeErrorEnvs.includes(global.cdkContextEnv),
  }: ErrorResponseOptions = {}
): APIGatewayProxyResult => {
  const bodyErrorObject: Record<string, unknown> = {
    statusCode,
    message: getReasonPhrase(statusCode),
    traceId,
  };

  if (description) {
    bodyErrorObject.description = description;
  }

  if (error instanceof Error) {
    console.error(`[${traceId}]`, error);

    if (exposeErrorMessage) {
      bodyErrorObject.error = String(error);
    }
  }

  return {
    statusCode,
    body: JSON.stringify({ error: bodyErrorObject }),
  };
};
