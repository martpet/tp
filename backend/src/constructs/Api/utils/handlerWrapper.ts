import { APIGatewayProxyHandlerV2 } from 'aws-lambda';

export const handlerWrapper =
  (func: APIGatewayProxyHandlerV2) =>
  (...params: Parameters<APIGatewayProxyHandlerV2>) =>
    func(...params);
