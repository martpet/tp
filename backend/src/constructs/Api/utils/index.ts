export * from './cookieName';
export * from './errorResponse';
export * from './getIdTokenPayload';
export * from './handlerWrapper';
export * from './itResolvesWithErrorResponse';
export * from './itResolvesWithLambdaEdgeErrorResponse';
export * from './lambdaEdgeErrorResponse';
export * from './parseEventCookies';
export * from './parseLambdaEdgeEventCookies';
