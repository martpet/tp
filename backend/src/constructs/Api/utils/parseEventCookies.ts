import { APIGatewayProxyEventV2 } from 'aws-lambda';
import cookie from 'cookie';

import { EndpointsCookies } from '../types';

export const parseEventCookies = <T extends keyof EndpointsCookies>(
  event: APIGatewayProxyEventV2
): Partial<Record<EndpointsCookies[T][number], string>> => {
  try {
    const cookieHeaderString = event.cookies?.join(';');
    if (!cookieHeaderString) throw new Error();
    return cookie.parse(cookieHeaderString);
  } catch (e) {
    return {};
  }
};
