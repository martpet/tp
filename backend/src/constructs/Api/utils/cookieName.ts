import { ValueOf } from 'type-fest';

import { EndpointsCookies } from '~/constructs/Api/types';

export const cookieName = (name: ValueOf<EndpointsCookies>[number]) => name;
