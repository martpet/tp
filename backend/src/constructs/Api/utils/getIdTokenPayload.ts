import camelcaseKeys from 'camelcase-keys';

import { IdTokenPayload } from '~/constructs/Api/types';

export const getIdTokenPayload = (idToken: string): IdTokenPayload => {
  const payload = JSON.parse(Buffer.from(idToken.split('.')[1], 'base64').toString());

  return camelcaseKeys(payload);
};
