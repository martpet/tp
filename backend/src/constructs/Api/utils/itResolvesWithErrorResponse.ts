import { APIGatewayProxyHandlerV2 } from 'aws-lambda';

import { errorResponse } from '~/constructs/Api/utils';

export const itResolvesWithErrorResponse = (
  handler: APIGatewayProxyHandlerV2,
  args: Parameters<APIGatewayProxyHandlerV2>
) => {
  it('calls "errorResponse" with correct args', async () => {
    await handler(...args);
    expect(vi.mocked(errorResponse).mock.calls).toMatchSnapshot();
  });

  it('resolves with a correct value', () => {
    return expect(handler(...args)).resolves.toMatchSnapshot();
  });
};
