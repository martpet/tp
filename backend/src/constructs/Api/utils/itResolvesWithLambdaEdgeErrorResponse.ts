import { LambdaEdgeViewerRequestHandler } from '~/constructs/Api/types';
import { lambdaEdgeErrorResponse } from '~/constructs/Api/utils';

export const itResolvesWithLambdaEdgeErrorResponse = (
  handler: LambdaEdgeViewerRequestHandler,
  args: Parameters<LambdaEdgeViewerRequestHandler>
) => {
  it('calls "lambdaEdgeErrorResponse" with correct args', async () => {
    await handler(...args);
    expect(vi.mocked(lambdaEdgeErrorResponse).mock.calls).toMatchSnapshot();
  });

  it('resolves with a correct value', () => {
    return expect(handler(...args)).resolves.toMatchSnapshot();
  });
};
