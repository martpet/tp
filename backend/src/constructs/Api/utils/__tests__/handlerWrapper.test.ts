import { APIGatewayProxyHandlerV2 } from 'aws-lambda';

import { handlerWrapper } from '~/constructs/Api/utils';

const fn = vi.fn().mockReturnValue('dummyFnReturnValue');
const fnArgs = ['dummyFnArg'] as unknown as Parameters<APIGatewayProxyHandlerV2>;

describe('"handlerWrapper" returns a wrapper function that', () => {
  const wrapperFn = handlerWrapper(fn);

  test('calls the wrapped function with its call args ', () => {
    wrapperFn(...fnArgs);
    expect(fn.mock.calls).toMatchSnapshot();
  });

  test('returns the return value of the wrapped function ', () => {
    expect(wrapperFn(...fnArgs)).toMatchSnapshot();
  });
});
