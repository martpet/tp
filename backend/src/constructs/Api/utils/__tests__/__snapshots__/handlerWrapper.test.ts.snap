// Vitest Snapshot v1

exports[`"handlerWrapper" returns a wrapper function that > calls the wrapped function with its call args  1`] = `
[
  [
    "dummyFnArg",
  ],
]
`;

exports[`"handlerWrapper" returns a wrapper function that > returns the return value of the wrapped function  1`] = `"dummyFnReturnValue"`;
